pkg/controller/factory/base_controller.go

```diff
@@ -0,0 +1,114 @@
+package factory
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+// baseController represents generic Kubernetes controller boiler-plate
+type baseController struct {
+	cachesToSync    []cache.InformerSynced
+	sync            func(ctx context.Context, controllerContext SyncContext) error
+	resyncEvery     time.Duration
+	ctx             SyncContext
+	shutdownContext context.Context
+}
+
+var _ Controller = &baseController{}
+
+func (c *baseController) Run(ctx context.Context, workers int) {
+	shutdownContext, shutdownComplete := context.WithCancel(context.Background())
+	if c.shutdownContext != nil {
+		panic(fmt.Sprintf("controller %q is already running", c.ctx.ControllerName()))
+	}
+	c.shutdownContext = shutdownContext
+
+	defer shutdownComplete()
+	defer utilruntime.HandleCrash()
+	defer c.ctx.Queue().ShutDown()
+	defer klog.Infof("Shutting down %s ...", c.ctx.ControllerName())
+
+	klog.Infof("Starting %s ...", c.ctx.ControllerName())
+	if !cache.WaitForCacheSync(ctx.Done(), c.cachesToSync...) {
+		return
+	}
+	klog.V(5).Infof("Caches synced for controller %s", c.ctx.ControllerName())
+
+	var workerWaitGroup sync.WaitGroup
+
+	for i := 1; i <= workers; i++ {
+		klog.Infof("Starting #%d worker of %s controller ...", i, c.ctx.ControllerName())
+		workerWaitGroup.Add(1)
+		go wait.UntilWithContext(ctx, func(ctx context.Context) {
+			defer workerWaitGroup.Done()
+			defer klog.Infof("Shutting down worker of %s controller ...", c.ctx.ControllerName())
+			c.runWorker(ctx)
+		}, time.Second)
+	}
+
+	// if periodical resync is requested, run it.
+	go c.runPeriodicalResync(ctx, c.resyncEvery)
+
+	// wait for controller shutdown to be requested
+	<-ctx.Done()
+
+	// wait for all workers to finish their jobs
+	workerWaitGroup.Wait()
+}
+
+func (c *baseController) Sync(ctx context.Context, syncCtx SyncContext) error {
+	return c.sync(ctx, syncCtx)
+}
+
+func (c *baseController) ShutdownContext() context.Context {
+	return c.shutdownContext
+}
+
+func (c *baseController) runPeriodicalResync(ctx context.Context, interval time.Duration) {
+	if interval == 0 {
+		return
+	}
+	go wait.UntilWithContext(ctx, func(ctx context.Context) {
+		if err := c.sync(ctx, c.ctx.(*controllerContext).withQueueObject(nil)); err != nil {
+			utilruntime.HandleError(fmt.Errorf("periodical resync of controller %s failed: %v", c.ctx.ControllerName(), err))
+		}
+	}, interval)
+}
+
+func (c *baseController) runWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
```

Comment:
don't we have some panic handling somewhere usually?

Reply:
in Run()